### DISCUSSION TITLE
`setuptools.find_packages()` without parameters not excluding `tests`

### DISCUSSION THREAD

**Discussion by penguinpee** [pypa/setuptools#4082] [Q&A] [answered]
Hi there,

It's me again with another `setuptools.find_packages()` mystery.

I'm trying to build a wheel for [RTFDE](https://github.com/seamustuohy/RTFDE) from source. Upstream uses [`packages=setuptools.find_packages()`](https://github.com/seamustuohy/RTFDE/blob/66780b8b1bc708f970de497e6e0c73d6f4a1a377/setup.py#L15) and none of the keywords (`packages`, `py_modules`, `ext_modules`) preventing setuptools automatic discovery from doing its magic are present.

When I run `python3 -m build --wheel --outdir foo/wheel`, the wheel contains the `tests/` directory and everything in it.

Options to work around it:

1. Remove empty `tests/__init__.pay`
2. Add `exclude=['tests*'],` to `setuptools.find_packages()`

According to the [Automatic Discovery](https://setuptools.pypa.io/en/latest/userguide/package_discovery.html#flat-layout) documentation neither should be necessary. What am I missing?

I tried this with setuptools `68.2.2` and `62.6.0`. Both failed in correctly discovering the module.

**Comment by abravalheri** [comment, marked as answer]
> According to the Automatic Discover should be necessary. What am I missing?

The docs suggest you leave out the `packages` config completely, so setuptools finds it out based on some heuristics (shold work for the majority of pckages). This paragraph from the docs is relevant:

> Automatic discovery will only be enabled if you don’t provide any configuration for `packages` and `py_modules`. If at least one of them is explicitly set, automatic discovery will not take place.

If automatic discovery is not enough for your project structure, then you can use `find_packages` for customisation. Note, however, that when you use `find_packages `, you are in complete control: no heuristics are applied. So you have to be explicit about what you want to include or exclude. With `find_packages` or `find_namespace_packages`.

It is usually a good idea to start with `include=["RTFDE*"]` and then trim the excesses with `exclude=`.

**Comment by abravalheri** [comment]
The reason why it contains `tests` is because that folder is a conventional Python package that is scanned as intended by `find_packages`.

If the original author does not mean to distribute `tests`, then the configuration for `packages=` needs to be changed.

**Comment by penguinpee** [comment]
> > According to the Automatic Discover should be necessary. What am I missing?
> 
> The docs suggest you leave out the `packages` config completely, so setuptools finds it out based on some heuristics (shold work for the majority of pckages). This paragraph from the docs is relevant:
> 
> > Automatic discovery will only be enabled if you don’t provide any configuration for `packages` and `py_modules`. If at least one of them is explicitly set, automatic discovery will not take place.

Thanks! That was it.

I seem to have a hard time remembering that leaving out `packages=setuptools.find_packages()` is a valid configuration and the wording in the documentation does not make that any easier.

I'll try to remember that. But maybe adding something like:


:warning: If you want automatic package discovery, ***don't use*** `packages_find()` in `$FILE`!

The above in addition to the info box that mentions not defining any of the other parameters. Or maybe on top of that box. Having `packages=` seems the intuitive thing to do.

**Comment by penguinpee** [comment]
> If the original author does not mean to distribute `tests`, then the configuration for `packages=` needs to be changed.

It's almost always a mistake. But one that happens very often. I can see why.
